test/src/main/java/org/apache/accumulo/test/fate/zookeeper/ZooLockIT.java

```diff
@@ -306,6 +365,262 @@ public void testUnexpectedEvent() throws Exception {
+      zl2.lock(zlw2, "test1".getBytes(UTF_8));
+      // The call above creates two nodes in ZK because of the overridden create method in
+      // ZooKeeperWrapper.
+      // The nodes created are:
+      // zlock#00000000-0000-0000-0000-BBBBBBBBBBBB#0000000002
+      // zlock#00000000-0000-0000-0000-BBBBBBBBBBBB#0000000003
+      //
+      // ZooLock should realize this and remove the latter one and place a watcher on the first one
+      // in case
+      // the ZooKeeper ephemeral node is deleted by some external process.
+      // Because zlock#00000000-0000-0000-0000-BBBBBBBBBBBB#0000000002 is not the first child in the
+      // list, it
+      // places a watcher on zlock#00000000-0000-0000-0000-AAAAAAAAAAAA#0000000000 so that it may
+      // try to acquire
+      // the lock when zlock#00000000-0000-0000-0000-AAAAAAAAAAAA#0000000000 is removed.
+
+      assertTrue(zlw1.isLockHeld());
+      assertFalse(zlw2.isLockHeld());
+
+      List<String> children = zk1.getChildren(parent, false);
+      assertTrue(children.contains("zlock#00000000-0000-0000-0000-AAAAAAAAAAAA#0000000000"));
+      assertFalse("this node should have been deleted",
+          children.contains("zlock#00000000-0000-0000-0000-AAAAAAAAAAAA#0000000001"));
+      assertTrue(children.contains("zlock#00000000-0000-0000-0000-BBBBBBBBBBBB#0000000002"));
+      assertFalse("this node should have been deleted",
+          children.contains("zlock#00000000-0000-0000-0000-BBBBBBBBBBBB#0000000003"));
+
+      assertNull(zl1.getWatching());
+      assertEquals("/zlretry/zlock#00000000-0000-0000-0000-AAAAAAAAAAAA#0000000000",
+          zl2.getWatching());
+
+      zl1.unlock();
+      assertFalse(zlw1.isLockHeld());
+      zk1.close();
+
+      while (!zlw2.isLockHeld()) {
+        LockSupport.parkNanos(50);
+      }
+
+      assertTrue(zlw2.isLockHeld());
+      zl2.unlock();
+      zk2.close();
+
+    }
+
+  }
+
+  static class LockWorker implements Runnable {
+
+    private static final Logger LOG = LoggerFactory.getLogger(LockWorker.class);
+
+    private final String parent;
+    private final String name;
+    private final CountDownLatch getLockLatch;
+    private final CountDownLatch lockCompletedLatch;
+    private final CountDownLatch unlockLatch = new CountDownLatch(1);
+    private final RetryLockWatcher lockWatcher = new RetryLockWatcher();
+    private Exception ex = null;
+
+    public LockWorker(final String parent, final String name, final CountDownLatch lockLatch,
+        final CountDownLatch lockCompletedLatch) {
+      this.parent = parent;
+      this.name = name;
+      this.getLockLatch = lockLatch;
+      this.lockCompletedLatch = lockCompletedLatch;
+    }
+
+    public void unlock() {
+      unlockLatch.countDown();
+    }
+
+    public boolean holdsLock() {
+      return lockWatcher.isLockHeld();
+    }
+
+    @Override
+    public void run() {
+      ConnectedWatcher watcher = new ConnectedWatcher();
+      try {
```

Comment:
Can this move inside try? in case it throws exception?

```suggestion
      try {
        ConnectedWatcher watcher = new ConnectedWatcher();
```

Reply:
added to test